Add some tests for gen1-3 abilities

SBird1337 opened the pull request:
## Description
Adds tests for the following abilities:

 * Damp ✅
 * Intimidate ❌ [Bug: #2619 ]
 * Speed Boost ✅
 * Volt Absorb ✅
 * Insomnia ✅
 * Stench ✅ [Missing: Multi Hit interaction due to RNG]
 * Drizzle ✅
 * Limber ✅
 * Sand Veil ✅
 * Water Absorb ✅
 * Oblivous ✅
 * Cloud Nine ✅

Tests currently failing due to issues with the test system:

 * Compound Eyes: Fails due to a statistical anomaly when checking against OHKO moves.

## **Discord contact info**
Karathan#1337

mrgriffin reviewed (changes requested):
This is awesome, thanks! :heart:

Some stylistic notes:
- I would try to rephrase the Intimidate tests to start with the word "Intimidate".
- `NONE_OF` should be more like `NONE_OF { x; }` rather than `NONE_OF(x);`.
- `PASSES_RANDOMLY` should occur before `GIVEN`. This one maybe we should add a check for, the reason is that `GIVEN` (and the other blocks) are not executed when determining the cost of a test for scheduling, so they will be erroneously counted as 1 battle.

But it's really good work :)

I'm curious about that Bulbapedia Speed Boost one. If I remember tomorrow maybe I can check what happens on Showdown.

LOuroboros commented:
Dropping my tests for `ability_intimidate.c`, in case you want to add them:
```c
SINGLE_BATTLE_TEST("PlayerIntimidateOnSwitchOut")
{
    GIVEN
    {
        PLAYER(SPECIES_EKANS) { Ability(ABILITY_INTIMIDATE); }
        OPPONENT(SPECIES_MAGIKARP);
    }
    WHEN
    {
        TURN {}
    }
    SCENE
    {
        ABILITY_POPUP(player, ABILITY_INTIMIDATE);
        ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, opponent);
        MESSAGE("Ekans's Intimidate\ncuts Foe Magikarp's ATTACK!");
    }
}

SINGLE_BATTLE_TEST("EnemyIntimidateOnSwitchOut")
{
    GIVEN
    {
        PLAYER(SPECIES_MAGIKARP);
        OPPONENT(SPECIES_EKANS) { Ability(ABILITY_INTIMIDATE); }
    }
    WHEN
    {
        TURN {}
    }
    SCENE
    {
        ABILITY_POPUP(opponent, ABILITY_INTIMIDATE);
        ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, player);
        MESSAGE("Foe Ekans's Intimidate\ncuts Magikarp's ATTACK!");
    }
}
```

SBird1337 commented:
I'll have to think about / integrate the more verbose intimidate tests. For now I've addressed @mrgriffin's remarks. Please don't merge before squashing / ping me so I can squash it myself.

DizzyEggg commented on test/ability_damp.c:
```diff
@@ -0,0 +1,56 @@
+// Bug: The ability popups don't show up.
+SINGLE_BATTLE_TEST("Damp prevents damage from aftermath")
+{
+    KNOWN_FAILING;
```
PASS #2645

DizzyEggg commented on test/ability_volt_absorb.c:
```diff
@@ -0,0 +1,88 @@
+// and Jolteon is not healed.
+SINGLE_BATTLE_TEST("Volt Absorb activates on status moves")
+{
+    KNOWN_FAILING;
```
PASS #2635

SBird1337 commented:
I've revisited Speed Boost on Showdown which acted as Ground Truth for some of the more weird edge cases. The test would also fail there, so I removed it and call it weird phrasing from Bulbapedia.

I've rebased and squashed my commits, ready for another review :)

mrgriffin commented on test/ability_cloud_nine.c:
```diff
@@ -0,0 +1,16 @@
+        TURN { MOVE(opponent, MOVE_SANDSTORM); }
+        TURN {}
+    } SCENE {
+        NONE_OF { HP_BAR(player); }
```
You might like to check that the sandstorm move and/or end-of-turn animation doesn't play?

SBird1337 replied:
I think if I'm reading the ability correctly on bulbapedia the animations should still play (and they do) - the effects are just negated.

mrgriffin reviewed (approved):
Looking good, thanks! :heart:
